### PR TITLE
Use generic_wstring on Windows in PathToString

### PR DESCRIPTION
**Pull request by geoffthemedio**
Hopefully addresses https://github.com/freeorion/freeorion/issues/2205

### CODE REVIEW

File: server/SaveLoad.cpp [hunk 1]
```diff
@@ -540,7 +540,7 @@ void LoadEmpireSaveGameData(const std::string& filename, std::map<int, SaveGameE
     try {
         fs::path path = FilenameToPath(filename);
         DebugLogger() << "LoadEmpireSaveGameData: filename: " << filename << " path:" << path;
-        fs::ifstream ifs(path, std::ios_base::binary);
+        fs::ifstream ifs(path.generic_path(), std::ios_base::binary);
```
Comment (dbenage-cx) [inline server/SaveLoad.cpp]: Would need to bump minimum boost version to 1.63 (note this is commented as experimental).

path.generic [first appeared](https://github.com/boostorg/filesystem/commit/6e92c9a8b293c4b94969dd309405c80b67dacfdd) in boost version 1.60, with the rename to generic_path in 1.63

~[make_preferred](https://www.boost.org/doc/libs/1_58_0/libs/filesystem/doc/reference.html#path-modifiers)?~ (Edit: would not work either)

Comment (geoffthemedio) [reply server/SaveLoad.cpp]: Maybe
````
#ifdef FREEORION_LINUX
        fs::ifstream ifs(path.native(), std::ios_base::binary);
#else
        fs::ifstream ifs(path, std::ios_base::binary);
#endif
````

Comment (dbenage-cx) [reply server/SaveLoad.cpp]: In this case path.native == path.string
If the path separator is suspected to be the issue, may want to convert to path.generic on client side before sending to server (most/all such functions on linux perform no conversion).
`IsExistingFile` could be used to validate at server.

**Comment by geoffthemedio** [comment]
Try that now?

**Comment by o01eg** [comment]
@geoffthemedio Currently I don't have a Windows OS.

Could someone build 0.4.8 branch with this patch on Windows and try to choose savegame?

**Comment by geoffthemedio** [comment]
> ... try to choose savegame?

Could you clarify what you mean by that?

**Comment by o01eg** [comment]
I mean to load it in the lobby.

**Comment by geoffthemedio** [comment]
In master + this patch, I can load a multiplayer save in the MP lobby, on Windows, with non-Latin characters in the path, while connected to a localhost server.

File: server/SaveLoad.cpp (hunk 1 above)
Comment (Vezzra) [reply server/SaveLoad.cpp]: As #2205 has been reported for the release branch, a fix for it needs to be cherry picked. However, on the release branch we still work with v8 of the SDK, meaning, IIRC, the min boost version there is 1.58 or 1.59. I really don't want to have to upgrade the SDK/boost versions for the release branch, so a patch for #2205 must not rely on newer boost versions.

**Comment by geoffthemedio** [comment]
Restored use of `generic_path()` when the local boost version supports it. Might need to switch to `#ifdef` though...

**Comment by o01eg** [comment]
Looks like server-side `SaveLoad.cpp` fix don't work. Could someone test this PR with Windows client 0.4.8 branch and my server?

**Comment by geoffthemedio** [comment]
"don't work" how?

It should have been redundant, since the client-side fix is still in place.

**Comment by dbenage-cx** [comment]
@o01eg The game seemed to load ok (with Win10), might check logs.  Am rebuilding linux for local test if there is other issue.

**Comment by dbenage-cx** [comment]
As far as testing, LGTM.

As it is fairly straight forward, could implement our own utility function to replace using generic_path.

**Review by Vezzra — CHANGES_REQUESTED** [review]
@geoffthemedio,

> Restored use of generic_path() when the local boost version supports it. Might need to switch to #ifdef though...

Yep, definitely. If I try to build this PR applied to the release branch I get build errors (because `generic_path` is an unknown symbol).

**Comment by Vezzra** [comment]
@o01eg,

> Could someone build 0.4.8 branch with this patch on Windows and try to choose savegame?

With this PR applied to the release branch, I can load savegames on your server. Without the PR applied I can't (trying to load a game just won't do anything, don't know what happens, can't check the logs on your server).

So, as far as I can tell, this PR fixes things at least insofar as loading a savegame cross-platform works with that patch, and doesn't without it.

**Comment by geoffthemedio** [comment]
> If I try to build this PR applied to the release branch I get build errors (because generic_path is an unknown symbol).

Switched to `#if` instead of `if`

**Comment by Vezzra** [comment]
Ok, builds fine now with the release branch.

Wanted to make sure and run another quick test with @o01eg's multiplayer server, but couldn't (someone else was logged into the multiplayer lobby, but didn't respond, so I couldn't try to load a game, and then I lost connection to the server and couldn't reconnect).

However, the final adjustment has been trivial enough, so I don't expect any issues.

I'm going to merge and cherry pick this, so it can go into todays test builds.

**Comment by o01eg** [comment]
Good. Tomorrow I will be with my PC and could publish fixed version.

> someone else was logged into the multiplayer lobby, but didn't respond, so I couldn't try to load a game

This bug didn't allow even choose save game in the lobby.

**Pull request merged**